### ISSUE TITLE
client loses connection with server during repeated file changes

### ISSUE THREAD

**Issue by sgoings** [Azure/draft#79]
Sometimes when I'm making rapid code changes, draft loses connection with draftd.

The client errors with:
```
E0531 15:44:35.761289   67206 portforward.go:332] an error occurred forwarding 63658 -> 44135: error forwarding port 44135 to pod draftd-2336841885-pfbkf_kube-system, uid : exit status 1: 2017/05/31 21:44:36 socat[25667] E connect(5, AF=2 127.0.0.1:44135, 16): Connection refused
Error: there was an error running 'draft up': there was an error while dialing the server: unexpected EOF
```

Watching the Kubernetes pods during `draft up` shows:

```
default   zeroed-ladybird-zeroed-ladybird-1047063997-h5cvw   0/1       Pending   0         0s
default   zeroed-ladybird-zeroed-ladybird-1115286856-l7pgp   1/1       Terminating   0         1m
default   zeroed-ladybird-zeroed-ladybird-1047063997-h5cvw   0/1       Pending   0         0s
default   zeroed-ladybird-zeroed-ladybird-1047063997-h5cvw   0/1       ContainerCreating   0         0s
default   zeroed-ladybird-zeroed-ladybird-1047063997-pkj8p   0/1       Pending   0         0s
default   zeroed-ladybird-zeroed-ladybird-1047063997-pkj8p   0/1       Pending   0         0s
default   zeroed-ladybird-zeroed-ladybird-1047063997-pkj8p   0/1       ContainerCreating   0         0s
default   zeroed-ladybird-zeroed-ladybird-1115286856-l7pgp   0/1       Terminating   0         1m
default   zeroed-ladybird-zeroed-ladybird-1115286856-l7pgp   0/1       Terminating   0         1m
default   zeroed-ladybird-zeroed-ladybird-1115286856-l7pgp   0/1       Terminating   0         1m
kube-system   draftd-2336841885-pfbkf   0/1       Error     12        17h
default   zeroed-ladybird-zeroed-ladybird-1047063997-h5cvw   0/1       Running   0         1s
kube-system   draftd-2336841885-pfbkf   0/1       CrashLoopBackOff   12        17h
default   zeroed-ladybird-zeroed-ladybird-1047063997-pkj8p   0/1       Running   0         3s
default   zeroed-ladybird-zeroed-ladybird-1047063997-h5cvw   1/1       Running   0         10s
default   zeroed-ladybird-zeroed-ladybird-1115286856-3f7mt   1/1       Terminating   0         2m
default   zeroed-ladybird-zeroed-ladybird-1047063997-pkj8p   1/1       Running   0         10s
default   zeroed-ladybird-zeroed-ladybird-1115286856-3f7mt   0/1       Terminating   0         2m
default   zeroed-ladybird-zeroed-ladybird-1115286856-3f7mt   0/1       Terminating   0         2m
default   zeroed-ladybird-zeroed-ladybird-1115286856-3f7mt   0/1       Terminating   0         2m
```

The error'd draft pod's logs were:

```
kubectl logs draftd-2336841885-pfbkf --previous -n kube-system
time="2017-05-31T21:46:56Z" level=info msg="server is now listening at tcp://0.0.0.0:44135"
time="2017-05-31T21:47:54Z" level=info msg="POST /apps/zeroed-ladybird"
panic: repeated read on failed websocket connection

goroutine 23 [running]:
panic(0x1299040, 0xc42045b0e0)
    /usr/local/go/src/runtime/panic.go:500 +0x1a1
github.com/Azure/draft/vendor/github.com/gorilla/websocket.(*Conn).NextReader(0xc420078000, 0x0, 0xc4206061e0, 0xab79a4, 0x180001, 0x0)
    /home/jenkins/go/src/github.com/Azure/draft/vendor/github.com/gorilla/websocket/conn.go:893 +0x16d
github.com/Azure/draft/vendor/github.com/gorilla/websocket.(*Conn).ReadMessage(0xc420078000, 0x8, 0xc4205fdba0, 0x1e, 0x1e, 0xed0c1312a, 0xc418638e62)
    /home/jenkins/go/src/github.com/Azure/draft/vendor/github.com/gorilla/websocket/conn.go:954 +0x2f
github.com/Azure/draft/api.pingClient(0xc420078000, 0xc42036b3e0)
    /home/jenkins/go/src/github.com/Azure/draft/api/server.go:515 +0x40
created by github.com/Azure/draft/api.serveWs
    /home/jenkins/go/src/github.com/Azure/draft/api/server.go:290 +0xa8f
```

**Comment by judexzhu** [comment]
I have the similar issue , "draft up" has no problem, but when I change the local file content , it lost connection 

```
Watching local files for changes...
E0612 04:27:19.943717    2715 portforward.go:332] an error occurred forwarding 44907 -> 44135: error forwarding port 44135 to pod dabe1d16ce7acf6974e7f20e12f81b42bd25228266126a718cb925e9455b2c02, uid : exit status 1: 2017/06/12 11:27:19 socat[31686] E connect(6, AF=2 127.0.0.1:44135, 16): Connection refused
Error: there was an error running 'draft up': there was an error while dialing the server: unexpected EOF

```

```
Watching local files for changes...
E0612 04:33:49.350722    2749 portforward.go:178] lost connection to pod
Error: there was an error running 'draft up': there was an error while dialing the server: dial tcp [::1]:45570: getsockopt: connection refused
```

and I like to know how to complete remove draft with everything it created 

thanks

**Comment by bacongobbler** [comment]
> and I like to know how to complete remove draft with everything it created

since draft deploys projects via helm, you can just run `helm delete --purge draft` to delete draft from the server. You can also invoke the same command for the applications you deployed to the cluster.

**Comment by bacongobbler** [comment]
This issue appears to be due to [`pingClient`](https://github.com/Azure/draft/blob/master/api/server.go#L525-L535) reading from a closed websocket connection. I'll see if I can fix that.

**Comment by bacongobbler** [comment]
related: #184

**Comment by bacongobbler** [comment]
closed via #237
